cpp/src/randomforest/randomforest_impl.cuh

```diff
@@ -189,53 +189,56 @@ void rfClassifier<T>::fit(const cumlHandle& user_handle, const T* input,
                           RandomForestMetaData<T, int>*& forest) {
   this->error_checking(input, labels, n_rows, n_cols, false);
 
+  const cumlHandle_impl& handle = user_handle.getImpl();
   int n_sampled_rows = this->rf_params.rows_sample * n_rows;
   int n_streams = this->rf_params.n_streams;
-
-  const cumlHandle_impl& handle = user_handle.getImpl();
-  cudaStream_t stream = user_handle.getStream();
-  cumlHandle local_handle[n_streams];
-  cudaStream_t local_stream[n_streams];
-  for (int i = 0; i < n_streams; i++) {
-    CUDA_CHECK(cudaStreamCreate(&local_stream[i]));
-    local_handle[i].setStream(local_stream[i]);
+  if (n_streams > handle.getNumInternalStreams()) {
```

Comment:
this is not very good for performance. I think cuml use allow algorithms to set the number of internal streams .

Reply:
This is a tricky one!

Ideally, we want all resources managed inside `cumlHandle` to have been allocated upfront. That way, it is easier for us to amortize their cost over the run. Now, if we expose a `setNumInternalStreams` method it will break this assumption. This also means that we'll have to sync across all the previous set of streams, before creating these new ones, which in itself can be of some overhead. This is the reason I updated the constructor of `cumlHandle` to accept this parameter during the initialization itself.

In this specific case, instead of printing warning sign and then clamping down the `n_streams` variable, we could throw as `ASSERT`, explicitly asking callers to initialize the `cumlHandle` with appropriate number of internal streams. What say @cjnolet @vishalmehta1991 ?

Reply:
In other words, something like this:
```cpp
const int nStreams = 8;
cumlHandle handle(nStreams);
....
RF_params params;
params.n_streams = nStreams;
...
fit(..., params);
```

Reply:
Yeah makes sense. Assert is better for mismatch streams.

Reply:
Thanks. Done.